### PR TITLE
SAK-40018: user activity

### PR DESCRIPTION
**Pull request by plukasew**
https://jira.sakaiproject.org/browse/SAK-40018

See Jira ticket for details of this feature, including videos.

See USER_ACTIVITY_README.md in the PR for implementation notes.

### CODE REVIEW

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/EventRegistryServiceImpl.java [hunk 1]
```diff
@@ -344,21 +351,37 @@ public String getToolIcon(String toolId) {
...
+
+		return eventIdEventMap;
+	}
+
+	private void buildEventIdMaps()	{
+		eventIdToolMap = new HashMap<>();
+		eventIdEventMap = new HashMap<>();
+		Iterator<ToolInfo> i = getMergedEventRegistry().iterator();
+		while (i.hasNext()){
+			ToolInfo t = i.next();
```
Comment (adrianfish) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/EventRegistryServiceImpl.java]: use a for loop?

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/EventRegistryServiceImpl.java [hunk 2]
```diff
@@ -344,21 +351,37 @@ public String getToolIcon(String toolId) {
...
+			Iterator<EventInfo> iE = t.getEvents().iterator();
+			while(iE.hasNext()){
+				EventInfo e = iE.next();
```
Comment (adrianfish) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/EventRegistryServiceImpl.java]: use a for loop?

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/AnnouncementReferenceResolver.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	/**
+	 * Resolves the given event reference into meaningful details
+	 * @param eventRef the event reference
+	 * @param tips tips for parsing out the components of the reference
+	 * @param anncServ the announcement service
+	 * @return an AnnouncementData object, or ResolvedEventData.ERROR/PERM_ERROR
+	 */
+	public static ResolvedEventData resolveReference( String eventRef, List<EventParserTip> tips, AnnouncementService anncServ )
+	{
+        if( StringUtils.isBlank( eventRef ) || anncServ == null )
```
Comment (adrianfish) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/AnnouncementReferenceResolver.java]: Indentation

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/AnnouncementReferenceResolver.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            log.warn( "Cannot resolve reference. Reference is null/empty or service(s) are not initialized." );
+            return ResolvedEventData.ERROR;
+        }
+
+		GenericRefParser.GenericEventRef ref = GenericRefParser.parse( eventRef, tips );
+        String channelRef = "/announcement/channel/" + ref.contextId + "/" + ref.subContextId;
+		try
```
Comment (adrianfish) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/AnnouncementReferenceResolver.java]: Indentation

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/AnnouncementReferenceResolver.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			log.warn( "Unable to retrieve channel/message by ref/id.", iue );
+		}
+		catch( PermissionException pe )
+		{
+			log.warn( "Permission exception trying to retrieve channel/message.", pe );
+			return ResolvedEventData.PERM_ERROR;
+		}
+
+		log.warn( "Unable to retrieve data; ref = " + eventRef );
+        return ResolvedEventData.ERROR;
```
Comment (adrianfish) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/AnnouncementReferenceResolver.java]: Indentation

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/CalendarReferenceResolver.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+	/**
+	 * Resolves the given reference into meaningful details
+	 * @param eventType the event type
+	 * @param eventRef the event reference
+	 * @param tips tips for parsing out the components of the reference
+	 * @param calServ the calendar service
+	 * @return a CalendarEntryData object, or ResolvedEventData.ERROR/PERM_ERROR/NO_DETAILS
+	 */
+    public static ResolvedEventData resolveReference( final String eventType, final String eventRef,
```
Comment (adrianfish) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/CalendarReferenceResolver.java]: Indentation

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/CalendarReferenceResolver.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+			List<EventParserTip> tips, CalendarService calServ )
+    {
+        if( StringUtils.isBlank( eventRef ) || calServ == null )
+        {
+            log.warn( "Cannot resolve reference. Reference is null/empty or service(s) are not initialized." );
+            return ResolvedEventData.ERROR;
+        }
+
+		GenericRefParser.GenericEventRef ref = CalendarRefParser.parse( eventType, eventRef, tips );
+        String calendarRef = "/calendar/calendar/" + ref.contextId + "/";
```
Comment (adrianfish) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/event/detailed/refresolvers/CalendarReferenceResolver.java]: Indentation

**Comment by bjones86** [comment]
I've fixed a few of the Codacy complaints here. The rest I feel are unnecessary and won't be addressed.

**Comment by bjones86** [comment]
Please don't merge this until I've squashed the commits into the first one(Paul's). Subsequently, it will need to be merged with 'Rebase and merge' after I've squashed, so that Paul retains authorship credit.

**Comment by bjones86** [comment]
@adrianfish, thanks for reviewing this. I've addressed the issues you brought up around indentation, lombok, and a few other minor suggestions.

> Seeing as a lot of these are new classes I'd like to see them adhere to the Google spec for Java formatting. Two spaces for tabs make deeply nested code a lot easier to read ,and set us up well for a more functional future. Opening braces at the end of the definition compress the code with detracting from readability (I used to be a brace on a new line man, but saw the light).

As far as I'm aware, there has never been a community consensus on adopting Google's Java formatting specifications (or any single coding style for that matter). I understand that you're trying to standardize things, however we've been down this path before (I once tried to coerce the community into adopting a coding style in the past...), and unfortunately it is extremely difficult to get everyone on the same page, and agreeing on one coding style -- no matter what style that may be. Personally, I'm not a fan of the Google style, for many reasons which I won't expand upon here....

In the past when these issues have come up, the community generally agreed on the following principals:

1. Respect the coding style of the file you're working in, if it's pre-existing. If the file uses a certain style, you should continue to use that style regardless of how crazy or bad you think it is.
2. If creating a new file, it's up to the discretion of the original developer to determine the coding style she/he prefers (provided it's not insane, and generally acceptable).

I'm all for adopting a coding style, but we need to have that conversation amongst the community before enforcing someone's preference. In the past, the community also agreed that we (unfortunately) missed our opportunity to reformat the entire code base into one unified coding style, which was when we transitioned from SVN to GitHub. Doing so after the fact risks muddying up the Git history, making it substantially more difficult to track down specific changes.

**Comment by adrianfish** [comment]
@bjones86 I get the git history thing - I'd never push for a wholesale reformatting of a codebase on that basis. I've just been seeing a fair few commits recently with 2 spaces and I think it makes the code a lot more readable - it also conforms with the latest javascript conventions which help there too. But, anyway, I'm not that fussed - just my opinion. Thanks for doing the lombok changes and sorting the indentation. I don't think anyone else wanted to do the review :)

**Comment by adrianfish** [comment]
More on the code formatting thing - the conventions of the people most likely to be supporting the code being added, in future, should also be considered. It's hard to reformat code later, for the reasons you cite, so code formatted in ways divergent from the current standards just ends up adding cognitive load. Sun/Oracle and Google advocate and have always advocated opening braces on the same line as the definition, for instance, and various brand new classes in this PR don't adhere to that. I'd say the majority of Java coders use either the Sun or Google standard. Also, most of the Sitestats code actually uses one of those conventions.

**Pull request merged**